### PR TITLE
minimally functioning sam parser

### PR DESCRIPTION
**Pull request by Koeng101**
This PR adds a sam file parser. Minimal functioning parser created, needs testing. Doing that next.

Not ready for merge.

### CODE REVIEW

File: io/sam/sam.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+			header.PG = append(header.PG, genericMap)
+		default:
+			return parser, Header{}, fmt.Errorf("Line %d should contain @SQ, @RG, @PG or @CO as top level tags, but they weren't found. Line text: %s", parser.line, line)
+		}
+	}
+	return parser, header, nil
+}
+
+// ParseNext parsers the next read from a parser. Returns an error upon EOF.
+func (parser *Parser) ParseNext() (Alignment, error) {
```
Comment (carreter) [inline io/sam/sam.go:151]: Please add some whitespace/comments to this to make it easier to read!

**Review by carreter — COMMENTED** [review]
Oops, probably left more comments than I should have... :sweat_smile:

File: io/sam/sam.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+const DefaultMaxLineSize int = 32768 // // 32kB is a magic number often used by the Go stdlib for parsing. We multiply it by two.
+
+// Each header in a SAM file begins with a @ followed by a two letter record
+// code type. Each line is tab delimited, and contains TAG:VALUE pairs. HD, the
+// first line, only occurs once, while SQ, RG, and PG can appear multiple
+// times. Finally, @CO contains user generated comments.
+//
+// For more information, check section 1.3 of the reference document.
+type Header struct {
+	HD map[string]string   // File-level metadata. Optional. If present, there must be only one @HD line and it must be the first line of the file.
```
[the remark this answers is not in the document]
Comment (Koeng101) [reply io/sam/sam.go:22]: Good idea. Lemme make a blank function for this

File: io/sam/sam.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
+package sam
```
[the remark this answers is not in the document]
Comment (Koeng101) [reply io/sam/sam.go:1]: Added, though gonna add a lot more descriptive once ready for merge.

File: io/sam/sam.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+// ParseNext parsers the next read from a parser. Returns an error upon EOF.
+func (parser *Parser) ParseNext() (Alignment, error) {
+	var alignment Alignment
+	lineBytes, err := parser.reader.ReadSlice('\n')
+	if err != nil {
+		return Alignment{}, err
+	}
+	parser.line++
+	line := strings.TrimSpace(string(lineBytes))
```
[the remark this answers is not in the document]
Comment (Koeng101) [reply io/sam/sam.go:158]: What does that mean exactly? What should be disambiguated?

File: io/sam/sam.go (hunk 1 above)
Comment (Koeng101) [reply io/sam/sam.go]: Ah yes, I definitely need to do that. Lil hard to read right now.

File: io/sam/sam.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (carreter) [reply io/sam/sam.go:179]: `parser.line` refers to the line number, while `line` refers to the line contents. IMO would help readability to make the difference explicit in the names.

**Comment by carreter** [comment]
What's the status of this? If it's ready for review, I'll take a look in the coming couple of days.

**Comment by Koeng101** [comment]
Well I haven't updated it to merge into #339 , so a bit to go there. I think it basically works though.

**Comment by carreter** [comment]
Alright, will give it a looksie tomorrow then!

**Pull request closed without merging**